### DISCUSSION TITLE
[Question] Data from legacy system

### DISCUSSION THREAD

**Discussion by sebastiannordby** [kgrzybek/modular-monolith-with-ddd#102] [General] [answered]
First of all, i'm very grateful for your contribution, i also follow your blog and love to read your concepts and ideas, @kgrzybek, also big thanks to contributors.

I'm currently trying to mockup the different modules in my domain, to see if i can structure a modular monolith, in a fine manner. 
This "issue" may contain many questions that may be hard to answer, but if you have time i would love to hear your opinion. I will try to simplify as much as possible. _I'm fairly new with designing these kind of systems._

In your _key assumptions_ you wrote:
5) **Each Module has it's own data in a separate schema - shared data is not allowed.**

In my domain i have the following modules:
- WMS(Warehouse Management)
- SalesOrder(Contains salesorderline, etc..)

This is a **multi tenancy** application, which means the database contains many companies, **each company** has it own set's of data. There is alot of crud, but also a big portion is operational.

_The data currently comes from a legacy-system, how would you go about making an integration here? There will be an integration running a time intervall and sending data to the API, writing both up and down(keeping in sync). So basically i'm asking how to import existing data into the system and should not this problem be a distributed transaction? There may also be data imports from other systems. The reason the data needs to be kept in sync is because the legacy have to live for a longer while. 

**How to implement an anti-corruption-layer from legacy system with god models, and the integration have to go both ways..**

Thanks in advance.

**Comment by kgrzybek** [comment]
Hi @sebastiannordby ,

I don't get a synchronization issue. You want to synchronize your new system and legacy system both ways? I mean if you change something in a new system then this change should be applied to a legacy system as well? Or you talking about data synchronization between modules (WMS, SalesOrder) in new system. Some high level diagram will be helpful :)

**Comment by sebastiannordby** [comment]
> Hi @sebastiannordby ,
> 
> I don't get a synchronization issue. You want to synchronize your new system and legacy system both ways? I mean if you change something in a new system then this change should be applied to a legacy system as well? Or you talking about data synchronization between modules (WMS, SalesOrder) in new system. Some high level diagram will be helpful :)

Yes, i want to synchronize the new system with the legacy. Since the legacy is a monolith with god models, it would need to get a DTO with all fields when pulling down from the new system. This goes the same way then pushing data, one large DTO with all fields to the new system, which would have to be distributed among the modules.

Please excuse my bad diagram skills.

![ModularMonolith](https://user-images.githubusercontent.com/24465003/91005199-97288a80-e5d6-11ea-8fe2-a8dc37834c84.png)

**Comment by kgrzybek** [comment, marked as answer]
The problem here is you don't have a master (single source of truth). As I understand one entity can be modified both in Legacy and New System. I see 3 solutions:
1) Distributed transaction - you need to update 2 databases in the same transaction. It will give you strong consistency, but you will have performance issues because of db locks
2) No distributed transaction - you update one database in one transaction and update the second database in another. This way you will have eventual consistency and you need handle conflicts (two edits in the same time for example). In your diagram I see cron interval, but I would say it is not an optimal solution. To have a minimal number of conflicts, the delay is important so you you should send the change (event/message) as quick as possible. You can use CDC (https://en.wikipedia.org/wiki/Change_data_capture) for this purpose for example.
3) You can apply Strangler Pattern and gradually migrate from legacy to new application. You need to intercept somehow requests to old Legacy Monolith and direct them to the new one. Of course this is not always possible. You can find more information on Martin Fowler's page https://martinfowler.com/bliki/StranglerFigApplication.html and in my article https://www.kamilgrzybek.com/design/strangling-net-framework-app-to-net-core/
